### ISSUE TITLE
Training loss becomes NAN during pretraining

### ISSUE THREAD

**Issue by larel5000** [HazyResearch/hyena-dna#68]
Hello,
I’m trying to do pretraning on my own data. But after some number of steps (1k or more), the loss becomes NAN. I tried to reduce the learning rate from 6e-4 to 5e-4, 1e-4, 6e-5, but it only seems to delay the problem.  My questions are :
- Did you encounter the same problem during your work ?
- Is gradient clipping implemented ? In my configuration, it’s supposed to be set to 1
- Is there a part of the code where I can track the gradient norm to see if it’s vanishing/exploding ?
- Can it be a float precision issue ? The precision is 16 in my case

Thank you.

Here is my configuration file :

I try to pre-process my own data. But after a certain number
epochs, the loss becomes NAN. I've tried reducing the learning rate from
from 6e-4 to 5e-4, 1e-4, 6e-5, but this only seems to delay the problem.
problem.  My questions are as follows:

CONFIG
├── train
│   └── seed: 2222                                                                                                                                           
│       interval: step                                                                                                                                       
│       monitor: test/loss                                                                                                                                   
│       mode: min                                                                                                                                            
│       ema: 0.0                                                                                                                                             
│       test: false                                                                                                                                          
│       debug: false                                                                                                                                         
│       ignore_warnings: false                                                                                                                               
│       state:                                                                                                                                               
│         mode: null                                                                                                                                         
│         n_context: 0                                                                                                                                       
│         n_context_eval: 0                                                                                                                                  
│       ckpt: null                                                                                                                                           
│       disable_dataset: false                                                                                                                               
│       validate_at_start: false                                                                                                                             
│       pretrained_model_path: null                                                                                                                          
│       pretrained_model_strict_load: true                                                                                                                   
│       pretrained_model_state_hook:                                                                                                                         
│         _name_: null                                                                                                                                       
│       post_init_hook:                                                                                                                                      
│         _name_: null                                                                                                                                       
│       layer_decay:                                                                                                                                         
│         _name_: null                                                                                                                                       
│         decay: 0.7                                                                                                                                         
│       gpu_mem: 49                                                                                                                                          
│       global_batch_size: 256                                                                                                                               
│                                                                                                                                                            
├── tolerance
│   └── logdir: ./resume                                                                                                                                     
│       id: null                                                                                                                                             
│                                                                                                                                                            
├── wandb
│   └── project: dna                                                                                                                                         
│       group: ''                                                                                                                                            
│       job_type: training                                                                                                                             
│       mode: online                                                                                                                                         
│       name: my_data_32k_1                                                                                                                                 
│       save_dir: .                                                                                                                                          
│       id: my_data_32k_1                                                                                                                                   
│                                                                                                                                                            
├── trainer
│   └── _target_: pytorch_lightning.Trainer                                                                                                                  
│       devices: 4                                                                                                                                           
│       accelerator: gpu                                                                                                                                     
│       accumulate_grad_batches: 4                                                                                                                           
│       max_epochs: 400                                                                                                                                      
│       gradient_clip_val: 1.0                                                                                                                               
│       log_every_n_steps: 10                                                                                                                                
│       limit_train_batches: 1.0                                                                                                                             
│       limit_val_batches: 1.0                                                                                                                               
│       num_nodes: 1                                                                                                                                         
│       precision: 16                                                                                                                                        
│                                                                                                                                                            
├── loader
│   └── batch_size: 50                                                                                                                                       
│       num_workers: 4                                                                                                                                       
│       pin_memory: true                                                                                                                                     
│       drop_last: true                                                                                                                                      
│                                                                                                                                                            
├── dataset
│   └── _name_: my_data                                                                                                                                     
│       fasta_list_file: /ml/hyena-dna/data/my_data/paths.csv                                                                    
│       dataset_name: my_data                                                                                                                               
│       tokenizer_name: char                                                                                                                                 
│       cache_dir: null                                                                                                                                      
│       max_length: 32768                                                                                                                                    
│       add_eos: true                                                                                                                                        
│       batch_size: 16                                                                                                                                       
│       batch_size_eval: 32                                                                                                                                  
│       num_workers: 12                                                                                                                                      
│       shuffle: true                                                                                                                                        
│       pin_memory: true                                                                                                                                     
│       max_length_val: 32768                                                                                                                                
│       max_length_test: 32768                                                                                                                               
│       pad_max_length: null                                                                                                                                 
│       rc_aug: false                                                                                                                                        
│       use_fixed_len_val: false                                                                                                                             
│       replace_N_token: false                                                                                                                               
│       pad_interval: false                                                                                                                                  
│                                                                                                                                                            
├── optimizer
│   └── _name_: adamw                                                                                                                                        
│       lr: 0.0006                                                                                                                                           
│       weight_decay: 0.1                                                                                                                                    
│       betas:                                                                                                                                               
│       - 0.9                                                                                                                                                
│       - 0.999                                                                                                                                              
│                                                                                                                                                            
├── scheduler
│   └── _name_: cosine_warmup_timm                                                                                                                           
│       t_in_epochs: false                                                                                                                                   
│       t_initial: 48000                                                                                                                                     
│       lr_min: 5.9999999999999995e-05                                                                                                                       
│       warmup_lr_init: 1.0e-06                                                                                                                              
│       warmup_t: 480.0                                                                                                                                      
│                                                                                                                                                            
├── callbacks
│   └── learning_rate_monitor:                                                                                                                               
│         logging_interval: step                                                                                                                             
│       timer:                                                                                                                                               
│         step: true                                                                                                                                         
│         inter_step: false                                                                                                                                  
│         epoch: true                                                                                                                                        
│         val: true                                                                                                                                          
│       params:                                                                                                                                              
│         total: true                                                                                                                                        
│         trainable: true                                                                                                                                    
│         fixed: true                                                                                                                                        
│       model_checkpoint:                                                                                                                                    
│         monitor: test/loss                                                                                                                                 
│         mode: min                                                                                                                                          
│         save_top_k: 1                                                                                                                                      
│         save_last: true                                                                                                                                    
│         dirpath: checkpoints/                                                                                                                              
│         filename: test/loss                                                                                                                                
│         auto_insert_metric_name: false                                                                                                                     
│         verbose: true                                                                                                                                      
│                                                                                                                                                            
├── task
│   └── _name_: lm                                                                                                                                           
│       loss: cross_entropy                                                                                                                                  
│       torchmetrics:                                                                                                                                        
│       - perplexity                                                                                                                                         
│       - num_tokens                                                                                                                                         
│                                                                                                                                                            
├── encoder
│   └── None                                                                                                                                                 
├── decoder
│   └── None                                                                                                                                                 
└── model
    └── _name_: lm                                                                                                                                           
        d_model: 256                                                                                                                                         
        n_layer: 4                                                                                                                                           
        d_inner: 1024                                                                                                                                        
        vocab_size: 12                                                                                                                                       
        resid_dropout: 0.0                                                                                                                                   
        embed_dropout: 0.1                                                                                                                                   
        fused_mlp: false                                                                                                                                     
        fused_dropout_add_ln: false                                                                                                                          
        checkpoint_mixer: true                                                                                                                               
        checkpoint_mlp: true                                                                                                                                 
        residual_in_fp32: true                                                                                                                               
        pad_vocab_size_multiple: 8                                                                                                                           
        layer:                                                                                                                                               
          _name_: hyena                                                                                                                                      
          emb_dim: 5                                                                                                                                         
          filter_order: 64                                                                                                                                   
          short_filter_order: 3                                                                                                                              
          l_max: 32770                                                                                                                                       
          modulate: true                                                                                                                                     
          w: 10                                                                                                                                              
          lr: 0.0006                                                                                                                                         
          wd: 0.0                                                                                                                                            
          lr_pos_emb: 0.0

**Comment by larel5000** [comment]
seems the t_initial was too high, so learning rate decrease too slowly. And also I hadn't enough sample in my training set.

**Comment by SkyWuUCSD** [comment]
so your end fix was decreasing t_initial?

**Comment by JiaLonghao1997** [comment]
> seems the t_initial was too high, so learning rate decrease too slowly. And also I hadn't enough sample in my training set.

I tried to train a model on my own dataset and encountered a similar problem. Below is my config.
```python
CONFIG
├── train
│   └── seed: 2222                                                                                                                              
│       interval: step                                                                                                                          
│       monitor: test/loss                                                                                                                      
│       mode: min                                                                                                                               
│       ema: 0.0                                                                                                                                
│       test: false                                                                                                                             
│       debug: false                                                                                                                            
│       ignore_warnings: false                                                                                                                  
│       state:                                                                                                                                  
│         mode: null                                                                                                                            
│         n_context: 0                                                                                                                          
│         n_context_eval: 0                                                                                                                     
│       ckpt: null                                                                                                                              
│       disable_dataset: false                                                                                                                  
│       validate_at_start: false                                                                                                                
│       pretrained_model_path: null                                                                                                             
│       pretrained_model_strict_load: true                                                                                                      
│       pretrained_model_state_hook:                                                                                                            
│         _name_: null                                                                                                                          
│       post_init_hook:                                                                                                                         
│         _name_: null                                                                                                                          
│       layer_decay:                                                                                                                            
│         _name_: null                                                                                                                          
│         decay: 0.7                                                                                                                            
│       gpu_mem: 41                                                                                                                             
│       global_batch_size: 32                                                                                                                   
│                                                                                                                                               
├── tolerance
│   └── logdir: ./resume                                                                                                                        
│       id: null                                                                                                                                
│                                                                                                                                               
├── wandb
│   └── None                                                                                                                                    
├── trainer
│   └── _target_: pytorch_lightning.Trainer                                                                                                     
│       devices: 2                                                                                                                              
│       accelerator: gpu                                                                                                                        
│       accumulate_grad_batches: 1                                                                                                              
│       max_epochs: 3                                                                                                                           
│       gradient_clip_val: 1.0                                                                                                                  
│       log_every_n_steps: 10                                                                                                                   
│       limit_train_batches: 1.0                                                                                                                
│       limit_val_batches: 1.0                                                                                                                  
│       num_nodes: 1                                                                                                                            
│       precision: 16                                                                                                                           
│                                                                                                                                               
├── loader
│   └── batch_size: 50                                                                                                                          
│       num_workers: 4                                                                                                                          
│       pin_memory: true                                                                                                                        
│       drop_last: true                                                                                                                         
│                                                                                                                                               
├── dataset
│   └── _name_: hg38                                                                                                                            
│       bed_file: /home1/jialh/metaHiC/LLMs/GTDB/gtdb_genomes_reps_r220/gtdb_genomes_reps_r220.sample10k.split32k.bed                           
│       fasta_file: /home1/jialh/metaHiC/LLMs/GTDB/gtdb_genomes_reps_r220/gtdb_genomes_reps_r220.sample10k.fna                                  
│       dataset_name: hg38                                                                                                                      
│       tokenizer_name: char                                                                                                                    
│       cache_dir: null                                                                                                                         
│       max_length: 32768                                                                                                                       
│       add_eos: true                                                                                                                           
│       batch_size: 32                                                                                                                          
│       batch_size_eval: 64                                                                                                                     
│       num_workers: 16                                                                                                                         
│       shuffle: true                                                                                                                           
│       pin_memory: true                                                                                                                        
│       max_length_val: 32768                                                                                                                   
│       max_length_test: 32768                                                                                                                  
│       pad_max_length: null                                                                                                                    
│       rc_aug: false                                                                                                                           
│       use_fixed_len_val: false                                                                                                                
│       replace_N_token: false                                                                                                                  
│       pad_interval: false                                                                                                                     
│                                                                                                                                               
├── optimizer
│   └── _name_: adamw                                                                                                                           
│       lr: 0.0006                                                                                                                              
│       weight_decay: 0.1                                                                                                                       
│       betas:                                                                                                                                  
│       - 0.9                                                                                                                                   
│       - 0.999                                                                                                                                 
│                                                                                                                                               
├── scheduler
│   └── _name_: cosine_warmup_timm                                                                                                              
│       t_in_epochs: false                                                                                                                      
│       t_initial: 2862                                                                                                                         
│       lr_min: 5.9999999999999995e-05                                                                                                          
│       warmup_lr_init: 1.0e-06                                                                                                                 
│       warmup_t: 28.62                                                                                                                         
│                                                                                                                                               
├── callbacks
│   └── learning_rate_monitor:                                                                                                                  
│         logging_interval: step                                                                                                                
│       timer:                                                                                                                                  
│         step: true                                                                                                                            
│         inter_step: false                                                                                                                     
│         epoch: true                                                                                                                           
│         val: true                                                                                                                             
│       params:                                                                                                                                 
│         total: true                                                                                                                           
│         trainable: true                                                                                                                       
│         fixed: true                                                                                                                           
│       model_checkpoint:                                                                                                                       
│         monitor: test/loss                                                                                                                    
│         mode: min                                                                                                                             
│         save_top_k: 1                                                                                                                         
│         save_last: true                                                                                                                       
│         dirpath: checkpoints/                                                                                                                 
│         filename: test/loss                                                                                                                   
│         auto_insert_metric_name: false                                                                                                        
│         verbose: true                                                                                                                         
│                                                                                                                                               
├── task
│   └── _name_: lm                                                                                                                              
│       loss: cross_entropy                                                                                                                     
│       torchmetrics:                                                                                                                           
│       - perplexity                                                                                                                            
│       - num_tokens                                                                                                                            
│                                                                                                                                               
├── encoder
│   └── None                                                                                                                                    
├── decoder
│   └── None                                                                                                                                    
└── model
    └── _name_: lm                                                                                                                              
        d_model: 256                                                                                                                            
        n_layer: 4                                                                                                                              
        d_inner: 1024                                                                                                                           
        vocab_size: 12                                                                                                                          
        resid_dropout: 0.0                                                                                                                      
        embed_dropout: 0.1                                                                                                                      
        fused_mlp: false                                                                                                                        
        fused_dropout_add_ln: true                                                                                                              
        checkpoint_mixer: true                                                                                                                  
        checkpoint_mlp: true                                                                                                                    
        residual_in_fp32: true                                                                                                                  
        pad_vocab_size_multiple: 8                                                                                                              
        layer:                                                                                                                                  
          _name_: hyena                                                                                                                         
          emb_dim: 5                                                                                                                            
          filter_order: 64                                                                                                                      
          short_filter_order: 3                                                                                                                 
          l_max: 32770                                                                                                                          
          modulate: true                                                                                                                        
          w: 10                                                                                                                                 
          lr: 0.0006                                                                                                                            
          wd: 0.0                                                                                                                               
          lr_pos_emb: 0.0      
```
The running results are as follows:
```python
[2025-01-03 08:09:31,884][sequence_lightning][INFO] - Optimizer group 0 | 63 tensors | lr 0.0006 | weight_decay 0.1
[2025-01-03 08:09:31,890][sequence_lightning][INFO] - Optimizer group 1 | 32 tensors | lr 0.0006 | weight_decay 0.0

  | Name               | Type             | Params
--------------------------------------------------------
0 | model              | ConvLMHeadModel  | 3.3 M
1 | encoder            | Identity         | 0
2 | decoder            | Identity         | 0
3 | train_torchmetrics | MetricCollection | 0
4 | val_torchmetrics   | MetricCollection | 0
5 | test_torchmetrics  | MetricCollection | 0
--------------------------------------------------------
3.3 M     Trainable params
0         Non-trainable params
3.3 M     Total params
6.555     Total estimated model params size (MB)
Epoch 0:   0%|                                                                                   | 1/8253 [00:06<15:37:10,  6.81s/it, loss=2.82][2025-01-03 08:10:01,736][torch.nn.parallel.distributed][INFO] - Reducer buckets have been rebuilt in this iteration.
[2025-01-03 08:10:01,738][torch.nn.parallel.distributed][INFO] - Reducer buckets have been rebuilt in this iteration.
Epoch 0:  10%|████████▋                                                                          | 865/8253 [28:42<4:05:15,  1.99s/it, loss=nan]
```

**Comment by larel5000** [comment]
t_initial:           # The total number of steps for pretraining, depends on the training size, the global batch size and the number of epoch wanted
  warmup_t:      # Defines the number of warmup steps (the number of step the learning rate should increase before starting to decrease

It's weird that you have 8253 steps per epoch but t_initial is 2862. Also I think warmup_t should be an integer
Here is the documentation of their scheduler: https://timm.fast.ai/SGDR

Maybe also try to decrease your lr after setting correctly your parameters

**Comment by larel5000** [comment]
> so your end fix was decreasing t_initial?

yes, but it is specific to my case. I think a lower lr should have also worked. But I didn't want to do a lot of epochs.
